src/validators/ArrayValidator.ts

```diff
@@ -18,78 +18,80 @@ import { Result } from '../lib/Result';
 import type { ExpandSmallerTuples, Tuple, UnshiftTuple } from '../lib/util-types';
 import { BaseValidator } from './imports';
 
-export class ArrayValidator<T extends unknown[], I = T[number]> extends BaseValidator<T> {
+type makeArray<T> = T extends any[] ? T : T[];
+
+export class ArrayValidator<T, I = makeArray<T>[number], P extends unknown[] = makeArray<T>> extends BaseValidator<P> {
```

Comment:
Please add the TODO from the PR description here in the code as well with: `// TODO (v4): `